File: LeetCode-Java/Merge k Sorted Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    
    static class ListNodeSort implements Comparator<ListNode> {
 
		public int compare(ListNode one, ListNode two) {
			return one.val - two.val;
		}
	}
    public ListNode mergeKLists(List<ListNode> lists) {
        if(lists == null || lists.size() <= 0) return null;
        PriorityQueue<ListNode> record = new PriorityQueue<ListNode>(lists.size(),new ListNodeSort());
        ListNode result = null, current = null;
        int nullCount = 0;
        for(int i = 0;i < lists.size();i ++)
        {
            if(lists.get(i) == null)
                nullCount ++;
            else
            {
                ListNode tempNode = lists.get(i);
                record.offer(tempNode);
                lists.set(i, tempNode.next);
            }
        }
        while(nullCount < lists.size())
        {
            ListNode headNode = record.poll();
            if(result == null) result = headNode;
            if(current != null) current.next = headNode;
            current = headNode;
            if(headNode.next == null)
                nullCount ++;
            else
            {
                record.offer(headNode.next);
            }
            headNode.next = null;
        }
        return result;
    }
}

File: Leetcode-cpp/merge_k_sorted_lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        auto my_compare = [](ListNode* v1, ListNode*v2) { return v1 == nullptr || (v2 != nullptr && v1->val > v2->val); };
        priority_queue<ListNode*, vector<ListNode*>, decltype(my_compare)> pq(my_compare);
        ListNode* fake_head = new ListNode(-1);
        ListNode* current = fake_head;
        for (auto val : lists) {
            pq.push(val);
        }
        while (!pq.empty() && pq.top() != nullptr) {
            current->next = pq.top();
            current = current->next;
            ListNode* temp = pq.top()->next;
            pq.pop();
            pq.push(temp);
        }
        return fake_head->next;
    }
};